sdk/storage/azure-storage-file-share/azure/storage/fileshare/_download.py

```diff
@@ -380,12 +385,11 @@ def chunks(self):
             downloader=iter_downloader,
             chunk_size=self._config.max_chunk_get_size)
 
-    def readall(self):
-        # type: () -> bytes
+    def readall(self) -> Union[bytes, str]:
```

Comment:
Returning a `Union` here makes this method very unfriendly to use. TL;DR, we should just return bytes and add an ignore to the line that returns a `str`.

For a bit more context, this is probably here because it was copied from Blob. In Blob, `download_blob` supports a keyword called `encoding` which, if specified, means that we convert the response to a `str` using the specified encoding. This is bad and has caused lots of headaches in Blob. Because of this, you'll also noticed in Blob that the `StorageStreamDownloader` uses a `Generic` and this function will return just `T`, which will be `bytes` or `str` based on whether `encoding` was specified on the `download_blob` call.

Luckily, `download_file` does not officially document that is supports `encoding`. It seems to actually support it (again, probably because it was copied from Blob) but since it's not publicly supported, we can/should just assume it's not used/supported, and our types should reflect that. Hence why this should only return `bytes` even though it technically supports returning a `str`.

Reply:
Noted, thanks for the thorough walkthrough :D